File: leetcode/p00173/p00173.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        self.helper(root)

    def next(self) -> int:
        node = self.stack.pop()
        self.helper(node.right)
        return node.val

    def hasNext(self) -> bool:
        return self.stack
    
    def helper(self, node: Optional[TreeNode]):
        while node:
            self.stack.append(node)
            node = node.left

File: leetcode/p00173/p00173.go
package p00173

// type BSTIterator struct {
// 	stack  Stack
// 	memo   map[*TreeNode]bool
// 	cursor *TreeNode
// }

// func Constructor(root *TreeNode) BSTIterator {
// 	iterator := BSTIterator{
// 		stack:  CreateStack(),
// 		memo:   map[*TreeNode]bool{},
// 		cursor: nil,
// 	}

// 	iterator.stack.Put(root)
// 	for {
// 		node := iterator.stack.Pop()
// 		if !iterator.memo[node] {
// 			iterator.memo[node] = true
// 			if node.Right != nil {
// 				iterator.stack.Put(node.Right)
// 			}
// 			if node.Left != nil {
// 				iterator.stack.Put(node)
// 				iterator.stack.Put(node.Left)
// 			}
// 		}
// 		if node.Left == nil || iterator.memo[node.Left] {
// 			iterator.cursor = node
// 			break
// 		}
// 	}

// 	return iterator
// }

// func (this *BSTIterator) Next() int {
// 	prev := this.cursor

// 	this.cursor = nil
// 	for !this.stack.Empty() {
// 		node := this.stack.Pop()
// 		if !this.memo[node] {
// 			this.memo[node] = true
// 			if node.Right != nil {
// 				this.stack.Put(node.Right)
// 			}
// 			if node.Left != nil {
// 				this.stack.Put(node)
// 				this.stack.Put(node.Left)
// 			}
// 		}
// 		if node.Left == nil || this.memo[node.Left] {
// 			this.cursor = node
// 			break
// 		}
// 	}

// 	return prev.Val
// }

// func (this *BSTIterator) HasNext() bool {
// 	return this.cursor != nil
// }
